core/src/main/scala/org/apache/spark/deploy/master/ApplicationInfo.scala

```diff
@@ -43,6 +42,18 @@ private[spark] class ApplicationInfo(
   @transient var endTime: Long = _
   @transient var appSource: ApplicationSource = _
 
+  // A cap on the number of executors this application can have at any given time.
+  // By default, this is infinite. Only after the first allocation request is issued
+  // by the application will this be set to a finite value.
+  @transient var executorLimit: Int = _
+
+  // A set of workers on which this application cannot launch executors.
+  // This is used to handle kill requests when `spark.executor.cores` is NOT set. In this mode,
+  // at most one executor from this application can be run on each worker. When an executor is
+  // killed, its worker is added to the blacklist to avoid having the master immediately schedule
+  // a new executor on the worker.
+  @transient var blacklistedWorkers: mutable.HashSet[String] = _
```

Comment:
You should be able to do without *both* a mutable collection *and* a `var`. I'd make this a `val` and initialize it to an empty set right away. Nulls are evil, and I don't see any reason to live with a null.

Probably this `var` should not be public either.